### ISSUE TITLE
How do I resend a registration confirmation email

### ISSUE THREAD

**Issue by jlwhitfill** [aws/aws-aspnet-cognito-identity-provider#53]
So if I register a new user and for some reason fail to confirm the signup.  When I try to login later it fails with NOTALLOWED = TRUE, which means the email/account is not verified. Keep in mind I am using the email as the username for login.  Is there a way to resend the confirmation code email?  I cant seem to find anything.  When I try to run

`await _userManager.SendEmailConfirmationTokenAsync(<CognitoUser>)`

It fails with an OBJECT NOT SET TO INSTANCE OF REFERENCE and I am retrieving and passing a valid user.  I do do a down and dirty way of just changing the EMAIL attribute to the same email but Im not sure that forces a new email confirmation code to be sent.  This is important as I have the flag set to not allow the user to login if the email is not confirmed, so I need to be able to let the user choose to resend one if they didnt get it or they didnt see it.

**Comment by assyadh** [comment]
Can you please share the stacktrace? As well as examples of what you do with the email?

This is the cognito api call: https://github.com/aws/aws-aspnet-cognito-identity-provider/blob/ee9c9ae6d29131df683518df731ab2c8fe6af6e5/src/Amazon.AspNetCore.Identity.Cognito/CognitoUserStore/CognitoUserStore.cs#L341

It will send an email confirmation to the email that is on the email attribute of user in the Cognito backend.

**Comment by jlwhitfill** [comment]
So what I do, is on the login, I check to see if the result.IsNotAllowed = true, if it is true, I pass in the email address from the login and lookup the user by the email address:

`var user = await _userManager.FindByEmailAsync(Input.Email).ConfigureAwait(false);`

then I make the call to SendEmailConfirmation with that user

`var result2 = await _userManager.SendEmailConfirmationTokenAsync(user).ConfigureAwait(false);`

Once the second line of code runs I get this strack trace:

```
NullReferenceException: Object reference not set to an instance of an object.
Amazon.AspNetCore.Identity.Cognito.CognitoUserStore<TUser>.GetUserAttributeVerificationCodeAsync(TUser user, string attributeName, CancellationToken cancellationToken)
CognitoCoreAuth.Areas.Identity.Pages.Account.LoginModel.OnPostAsync(string returnUrl) in Login.cshtml.cs
+
                    var result2 = await _userManager.SendEmailConfirmationTokenAsync(user).ConfigureAwait(false);
Microsoft.AspNetCore.Mvc.RazorPages.Internal.ExecutorFactory+GenericTaskHandlerMethod.Convert<T>(object taskAsObject)
Microsoft.AspNetCore.Mvc.RazorPages.Internal.ExecutorFactory+GenericTaskHandlerMethod.Execute(object receiver, object[] arguments)
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.InvokeHandlerMethodAsync()
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.InvokeNextPageFilterAsync()
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
NToastNotify.NtoastNotifyAjaxToastsMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
Microsoft.AspNetCore.Builder.UseMiddlewareExtensions+<>c__DisplayClass5_1+<<UseMiddlewareInterface>b__1>d.MoveNext()
Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

The user is valid and the _userManager.FindByEmailAsync returns this user:

![image](https://user-images.githubusercontent.com/43940326/51947258-75a6f300-23e1-11e9-9dc5-1c3d48400e01.png)

At first I thought maybe I didnt initialize the _userManager, but I verified I am using it further up the code in another case scenario for changing the user's forgotten password and it works just fine there.

**Comment by jlwhitfill** [comment]
Anymore thoughts on this one?

**Comment by assyadh** [comment]
Hi @jlwhitfill I think I narrowed down the issue, it has to do with the fact that the cognito api expects the user to be authenticated to verify the attributes. I am trying to reproduce the issue with a cognito pool similar to yours.

**Comment by assyadh** [comment]
I have a PR in that adds an API for the purpose of resending the confirmation code for account signup.

In depth:

For Cognito, attribute verification is a different api than account signup verification. ie you can change your email and call the email verification api as you did, but this only verifies the email, not the account.

Now when calling the new api, you will get a confirmation code sent to your email/device, then need to call usermanager.ConfirmSignupAsync() with the token. This confirms the account AND the email/device at the same time.

Also note that you can call AdminConfirmSignupAsync() . This will confirm the account BUT NOT the email/device. This is when SendEmailConfirmationTokenAsync can later be used by a user with a confirmed account but not a confirmed email.

I will let you know when the new package is available on NuGet

**Comment by assyadh** [comment]
https://www.nuget.org/packages/Amazon.AspNetCore.Identity.Cognito/0.9.0.5  should fix the issue.

The usermanager now exposes ResendSignupConfirmationCodeAsync

**Comment by jlwhitfill** [comment]
This works correctly on my end, thank you!

**Comment by duaneking** [comment]
Do any plans exist to update the sample app, or at least document that SendEmailConfirmationTokenAsync  should not be used?

This is missing, and the use of the other SendEmailConfirmationTokenAsync  call that will always fail is somewhat misleading, IMHO.

I registered a user last night and got to account validation of it this morning to find the code I was sent last night no longer worked, as cognito considered the code I had been sent invalid, and any attempt to use SendEmailConfirmationTokenAsync method to send me a new code failed with an error from Cognito that was not helpful.

Hours of googling later I found this, and was able to use the ResendSignupConfirmationCodeAsync method mentioned above to send myself  a new validation token/code that I was able to use to validate the account using a validation endpoint on our side,  but that was a lot of pain that could have been fixed easily by simply not directing people to use SendEmailConfirmationTokenAsync  in the sample app if its not what we should be using.
